File: 0230/dfs.cpp
#include <cassert>
#include <iostream>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  int kthSmallest(TreeNode* root, int k) {
    int ret = 0;
    dfs(root, k, 0, ret);
    return ret;
  }

  int dfs(TreeNode* node, int k, int offset, int& ret) {
    if (!node) return 0;
    int left = dfs(node->left, k, offset, ret);
    int id = offset + left + 1;
    int right = dfs(node->right, k, id, ret);
    if (id == k) {
      ret = node->val;
    }
    return left + right + 1;
  }
};

int main() {
  Solution sol;
  TreeNode* node;

  //   3
  //  / \
  // 1   4
  //  \
  //   2
  node = CreateTreeNode("3 1 4 # 2");
  assert(sol.kthSmallest(node, 1) == 1);
  DeleteTreeNode(node);

  //       5
  //      / \
  //     3   6
  //    / \
  //   2   4
  //  /
  // 1
  node = CreateTreeNode("5 3 6 2 4 # # 1");
  assert(sol.kthSmallest(node, 3) == 3);
  DeleteTreeNode(node);

  //        10
  //      /    \
  //     4      12
  //    / \    /  \
  //   2   6  11  13
  //  / \ / \
  // 1  3 5  7
  node = CreateTreeNode("10 4 12 2 6 11 13 1 3 5 7");
  assert(sol.kthSmallest(node, 9) == 11);
  DeleteTreeNode(node);
}

File: 1530/dfs.py
from collections import defaultdict

from leetcode import TreeNode, String2Tree


class Solution:
    def __init__(self):
        self._ret = 0

    def countPairs(self, root: TreeNode, distance: int) -> int:
        self._ret = 0
        counts = self.dfs(root, distance)
        return self._ret

    def dfs(self, node, distance):
        if node is None:
            return {}
        # leaf node
        if node.left is None and node.right is None:
            return {1: 1}
        counts_left = self.dfs(node.left, distance)
        counts_right = self.dfs(node.right, distance)
        counts = defaultdict(int)
        for dist_left, count_left in counts_left.items():
            counts[dist_left+1] += count_left
        for dist_right, count_right in counts_right.items():
            counts[dist_right+1] += count_right
        for dist_left, count_left in counts_left.items():
            for dist_right, count_right in counts_right.items():
                dist = dist_left + dist_right
                if dist <= distance:
                    self._ret += count_left * count_right
        return counts


if __name__ == '__main__':
    sol = Solution()

    root, distance, expected = String2Tree("1 2 3 # 4"), 3, 1
    assert sol.countPairs(root, distance) == expected

    root, distance, expected = String2Tree("1 2 3 4 5 6 7"), 3, 2
    assert sol.countPairs(root, distance) == expected

    root, distance, expected = String2Tree("7 1 4 6 # 5 3 # # # # # 2"), 3, 1
    assert sol.countPairs(root, distance) == expected
